File: module_examples/vi/test_ffmpeg_demux.cpp
#include "module/vi/module_ffmpegDemux.hpp"

void output_callback(void* _ctx, shared_ptr<MediaBuffer> buffer)
{
    auto buffer_type = buffer->getMediaBufferType();

    if (buffer_type == BUFFER_TYPE_VIDEO) {
        auto v_buf = dynamic_pointer_cast<VideoBuffer>(buffer);
        if (v_buf) {
            ff_info("Video[%d] pts %ld, dts %ld, fd %d, data %p, bytes %ld\n",
                    v_buf->getMediaCodec(), v_buf->getPUstimestamp(),
                    v_buf->getDUstimestamp(), v_buf->getBufFd(),
                    v_buf->getActiveData(), v_buf->getActiveSize());
        }

        auto param = buffer->getImagePara();
        ff_info("Video foramt %s, width %d, height %d\n\n", v4l2GetFmtName(param.v4l2Fmt),
                param.width, param.height);
    } else if (buffer_type == BUFFER_TYPE_AUDIO) {
        ff_info("Audio[%d] pts %ld, dts %ld, data %p, bytes %ld\n", buffer->getMediaCodec(),
                buffer->getPUstimestamp(), buffer->getDUstimestamp(),
                buffer->getActiveData(), buffer->getActiveSize());

        auto param = buffer->getSamplePara();
        ff_info("Audio foramt %d, channels %d, sample_rate %d, nb_samples %d\n\n", param.fmt,
                param.channels, param.sample_rate, param.nb_samples);

    } else {
        ff_info("ETC[%d] pts %ld, dts %ld, data %p, bytes %ld\n\n", buffer->getMediaCodec(),
                buffer->getPUstimestamp(), buffer->getDUstimestamp(),
                buffer->getActiveData(), buffer->getActiveSize());
    }
}

void status_change_callback(void* ctx, ModuleStatus status)
{
    ff_info("Module state has changed(%d)\n", status);
}

int main(int argc, char** argv)
{
    int ret;

    if (argc < 2) {
        ff_error("\nUsage: %s rtsp://xxx\n", argv[0]);
        return -1;
    };

    // create FFmpeg demux module
    auto f_demux = make_shared<ModuleFFmpegDemux>(argv[1]);
    // f_demux->setInputFormat("rtsp");
    /* Reducing probesize can speed up the opening of streams.*/
    f_demux->setFormatOption("probesize", "200K", 0);
    /* Set the callback function for the demuxer */
    f_demux->setOutputDataCallback(nullptr, output_callback);
    f_demux->setStatusChangeCallback(nullptr, status_change_callback);
    ret = f_demux->init();
    if (ret < 0) {
        ff_error("Failed to init ffmpeg demux, %d\n", ret);
        return ret;
    }

    // Print the video stream information
    auto v_extra = f_demux->getExtraBuffer(BUFFER_TYPE_VIDEO);
    if (v_extra) {
        ff_info("Video: Extra codec %d , data %p, bytes %ld\n", v_extra->getMediaCodec(),
                v_extra->getActiveData(), v_extra->getActiveSize());

        auto param = v_extra->getImagePara();
        ff_info("foramt %s, width %d, height %d\n", v4l2GetFmtName(param.v4l2Fmt),
                param.width, param.height);
    }

    // Print the audio stream information
    auto a_extra = f_demux->getExtraBuffer(BUFFER_TYPE_AUDIO);
    if (a_extra) {
        ff_info("Audio: Extra codec %d , data %p, bytes %ld\n", a_extra->getMediaCodec(),
                a_extra->getActiveData(), a_extra->getActiveSize());
        auto param = a_extra->getSamplePara();
        ff_info("foramt %d, channels %d, sample_rate %d, nb_samples %d\n", param.fmt,
                param.channels, param.sample_rate, param.nb_samples);
    }

    ff_info("\n===========================================================================\n\n");

    // start
    f_demux->start();
    getchar();
    f_demux->stop();
    return 0;
}

File: module_examples/vi/test_ffmpeg_demux.py
#!/usr/bin/env python3

import sys
import ff_pymedia as ff

def output_callback(obj, buffer):
    buffer_type = buffer.getMediaBufferType()
    if buffer_type == ff.BUFFER_TYPE_VIDEO:
        v_param = buffer.getImagePara()
        print("Video[{}] format {}, width {}, height {}, pts {}, dts {}".format\
                (buffer.getMediaCodec(), v_param.v4l2Fmt, v_param.width,\
                v_param.height, buffer.getPUstimestamp(),\
                buffer.getDUstimestamp()))

    elif buffer_type == ff.BUFFER_TYPE_AUDIO:
        a_param = buffer.getSamplePara()
        print("Audio[{}] format {}, channels {}, sample rate {}, pts {}, dts {}"\
                .format(buffer.getMediaCodec(), a_param.fmt, a_param.channels,\
                a_param.sample_rate, buffer.getPUstimestamp(),\
                buffer.getDUstimestamp()))

    # get memoryview object
    data = buffer.getActiveData()
    print("Data size ", data.nbytes)

def status_change_callback(obj, status):
    print("Module state has changed: ", status)


def main():
    if len(sys.argv) > 1:
        file_path = sys.argv[1]
    else:
        file_path = input("Please enter the file path or url: ")

    # create FFmpeg demux module.
    input_source = ff.ModuleFFmpegDemux(file_path, -1)
    input_source.setFormatOption("probesize", "200K", 0);
    # Set the callback function for the demuxer.
    input_source.setOutputDataCallback(None, output_callback);
    input_source.setStatusChangeCallback(None, status_change_callback);
    ret = input_source.init();
    if ret < 0:
        print("Failed to init demuxer")
        return 1

    # View the video stream information.
    v_extra = input_source.getExtraBuffer(ff.BUFFER_TYPE_VIDEO);
    if v_extra:
        output_callback(None, v_extra)
    # View the audio stream information.
    a_extra = input_source.getExtraBuffer(ff.BUFFER_TYPE_AUDIO);
    if a_extra:
        output_callback(None, a_extra)

    print("\n============================START=============================\n")
    input_source.start()
    input("wait...")
    input_source.stop()
    print("\n============================STOP==============================\n")

if __name__ == "__main__":
    main()
